File: python/graphs/max-area-of-island.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        neighbors = [(0, 1), (-1, 0), (0, -1), (1, 0)]
        maxarea = 0

        def dfs(i, j):
            if i < 0 or j < 0 or i >= len(grid) or j >= len(grid[0]) or grid[i][j] != 1:
                return 0

            count = 1
            grid[i][j] = 2
            for neighbor in neighbors:
                count += dfs(i + neighbor[0], j + neighbor[1])

            return count

        for i in range(len(grid)):
            for j in range(len(grid[0])):
                maxarea = max(maxarea, dfs(i, j))

        return maxarea

File: c++/graphs/max-area-of-island.cc
class Solution {
public:
    int maxAreaOfIsland(vector<vector<int>>& grid) {
        std::vector<std::pair<int, int>> options = {std::make_pair(0, 1), std::make_pair(1, 0), std::make_pair(0, -1), std::make_pair(-1, 0)};
        int answer = 0;
        for (int i = 0; i < grid.size(); ++i) {
            for (int j = 0; j < grid[0].size(); ++j) {
                if (grid[i][j] != 1) {
                    continue;
                }
                answer = max(answer, dfs(i, j, grid, options));
            }
        }
        return answer;
    }

private:
    int dfs(int i, int j, std::vector<std::vector<int>>& grid, std::vector<std::pair<int, int>>& options) {
        if (i < 0 || j < 0 || i >= grid.size() || j >= grid[0].size() || grid[i][j] != 1) {
            return 0;
        }

        grid[i][j] = 0;

        int area = 1;
        for (auto [n_i, n_j] : options) {
            area += dfs(i+n_i, j+n_j, grid, options);
        }
        return area;
    }
};
